examples/cis3-nft-sponsored-txs/src/lib.rs

```diff
@@ -916,8 +869,19 @@ fn contract_permit<S: HasStateApi>(
         CustomContractError::Expired.into()
     );
 
+    let zeros: Vec<u8> = Vec::from([0, 0, 0, 0, 0, 0, 0, 0]);
+
     // Calculate the message hash.
-    let message_hash = crypto_primitives.hash_sha2_256(&to_bytes(&param.message)).0;
+    let message_hash = crypto_primitives
+        .hash_sha2_256(
+            &[
+                to_bytes(&param.signer),
+                zeros,
+                hex::encode(to_bytes(&param.message)).as_bytes().to_vec(),
+            ]
+            .concat(),
+        )
+        .0;
```

Comment:
It's rather annoying to have to do this hex encoding. I see it is necessary the way wallets are now.

This can also be done more efficiently I think without having to first deserialize the parameters and then serialize them back again here. It's a bit more complicated to write, but I think the gains will be worth it. I can try to write that when I'm in the office later today.

Reply:
I pushed a commit to this branch that does some of these optimizations.

I am not sure it is worth it as such, so feel free to revert it.

However one advantage is that we can now build the contract with `no-std`. With that the contract size is around 102kB.